File: eoin-mcdonnell/day-08/who-likes-it.js
//we need to add and for names one or more
// if there is three names the and need to be between the last two
// if there are more than three we only show the first two and add and x amount others like this
// we will need to add the names to the array
// we will need to for the last one be array.length - 2 others like this
//
/*-
function likes(names) {
  // TODO
  console.log(names)
  let namesArr = [];
  if(names === ""){
    return "no one likes this"
  } else {
    namesArr.push(names)
  }
  for (let i = 0; i < namesArr.length; i++){
    if(i ===  1) {
      return "${namesArr[i]} likes this"
    } else if (i === 2) {
      return "${namesArr[0]} and ${namesArr[1]} like this"
    } else if (i === 3){
      return "${namesArr[0]}, ${namesArr[1]}, and ${namesArr[3]} like this"
    } else if (i > 4) {
      return "${namesArr[0]} and ${namesArr[1]} and ${namesArr.length -2} other like this."
    }
  }
}
-*/

//names is already an array, i do not need to add them to an array


function likes(names) {
  if (names.length == 0) {
    return "no one likes this";
  } else if (names.length == 1) {
    return names[0] + " likes this";
  } else if (names.length == 2) {
    return names[0] + " and " + names[1] + " like this";
  } else if (names.length == 3) {
    return names[0] + ", " + names[1] + " and " + names[2] + " like this";
  } else {
    return names[0] + ", " + names[1] + " and " + (names.length - 2) + " others like this";
  }
}

/*
  Ember's Feedback:
  - Great job!
  - Put unused / old code in a notes.js file so the submission is kept tidy and you can reference your old code later
    if it is helpful for you to do that.
*/
// Alternative Solution
function likes(names) {
  switch (names.length) {
    case 0:
      return 'no one likes this';
    case 1:
      return `${names[0]} likes this`;
    case 2:
      return `${names[0]} and ${names[1]} like this`;
    case 3:
      return `${names[0]}, ${names[1]} and ${names[2]} like this`;
    default:
      return `${names[0]}, ${names[1]} and ${names.length - 2} others like this`;
  }
}

File: henry-knollenberg/day-38/who_likes_it.py
# if length 1, 2, 3, or else
#   1 - name0 likes this
#   2 - name0 and name1 like this
#   3 - name0, name1 and name2 like this
#   4 - name0, name1 and 2 others like this


def likes(names):
    if len(names) == 0:
        return "no one likes this"
    elif len(names) == 1:
        return f"{names[0]} likes this"
    elif len(names) == 2:
        return f"{names[0]} and {names[1]} like this"
    elif len(names) == 3:
        return f"{names[0]}, {names[1]} and {names[2]} like this"
    else:
        return f"{names[0]}, {names[1]} and {len(names) - 2} others like this"


"""
    Ember's Feedback:
    - Good job!
    - Python has so many helpful ways of doing this! How fun :D
"""


# Alternative Solutions
def likes(names):
    match len(names):
        case 0:
            return "no one likes this"
        case 1:
            return f"{names[0]} likes this"
        case 2:
            return f"{names[0]} and {names[1]} like this"
        case 3:
            return f"{names[0]}, {names[1]} and {names[2]} like this"
        case _:
            return f"{names[0]}, {names[1]} and {len(names) - 2} others like this"


# OR
def likes(names):
    n = len(names)
    return {
        0: "no one likes this",
        1: "{} likes this",
        2: "{} and {} like this",
        3: "{}, {} and {} like this",
        4: "{}, {} and {others} others like this",
    }[min(4, n)].format(*names[:3], others=n - 2)


# OR
def likes(names):
    match names:
        case []:
            return "no one likes this"
        case [a]:
            return f"{a} likes this"
        case [a, b]:
            return f"{a} and {b} like this"
        case [a, b, c]:
            return f"{a}, {b} and {c} like this"
        case [a, b, *rest]:
            return f"{a}, {b} and {len(rest)} others like this"
